File: javascript/yacht/yacht.js
export const score = (dice, category) => {
    const scoreFunctionMap = {
        'ones': (rolls) => numberCategories(1, rolls),
        'twos': (rolls) => numberCategories(2, rolls),
        'threes': (rolls) => numberCategories(3, rolls),
        'fours': (rolls) => numberCategories(4, rolls),
        'fives': (rolls) => numberCategories(5, rolls),
        'sixes': (rolls) => numberCategories(6, rolls),
        'full house': fullHouse,
        'four of a kind': fourOfAKind,
        'little straight': littleStraight,
        'big straight': bigStraight,
        'choice': choice,
        'yacht': yacht,
    };

    return scoreFunctionMap[category](dice);
};

const numberCategories = (number, dice) => number * freqNumber(dice, number);

const fullHouse = (dice) => isFullHouse(dice) 
            ? dice.reduce((a, b) => a + b, 0) : 0;

const fourOfAKind = (dice) => 
    4 * ((dice.find(number => freqNumber(dice, number) >= 4)) ?? 0);

const littleStraight = (dice) => 
    JSON.stringify(sortArray(dice)) === '[1,2,3,4,5]' ? 30 : 0;

const bigStraight = (dice) => 
    JSON.stringify(sortArray(dice)) === '[2,3,4,5,6]' ? 30 : 0;

const choice = (dice) => dice.reduce((a, b) => a + b);

const isFullHouse = (dice) => 
    dice.some(number => freqNumber(dice, number) === 3)
    && dice.some(number => freqNumber(dice, number) === 2);

const yacht = (dice) => 
    dice.some(number => freqNumber(dice, number) === 5) ? 50 : 0;

const freqNumber = (dice, number) => dice.filter(n => number === n).length;

const sortArray = (array) => array.sort((a, b) => a - b);

File: clojure/yacht/src/yacht.clj
(ns yacht)

(defn number-freq [number, xs]
  (cond 
    (empty? xs) 0
    (= (first xs) number) (+ 1 (number-freq number (rest xs)))
    :else (number-freq number (rest xs))
  )
)

(defn score-yacht [dice]
  (if (= (count (set dice)) 1)
    50
    0
  )
)

(defn score-n [dice, n]
  (* n (number-freq n, dice))
)

(defn score-full-house [dice]
  (def num-a (first (sort dice)))
  (def num-b (last (sort dice)))
  
  (cond
    (and (= (number-freq num-a dice) 2) (= (number-freq num-b dice) 3)) (reduce + dice)
    (and (= (number-freq num-a dice) 3) (= (number-freq num-b dice) 2)) (reduce + dice)
    :else 0
  )
)

(defn score-foak [dice]
  (def num-a (first (sort dice)))
  (def num-b (last (sort dice)))
  
  (cond
    (>= (number-freq num-a dice) 4) (* 4 num-a)
    (>= (number-freq num-b dice) 4) (* 4 num-b)
    :else 0
  )
)

(defn score-little-straight [dice]
  (if (= [1 2 3 4 5] (sort dice))
    30
    0
  )
)

(defn score-big-straight [dice]
  (if (= [2 3 4 5 6] (sort dice))
    30
    0
  )
)

(defn score [dice, category]
  (case category
    "yacht" (score-yacht dice)
    "ones" (score-n dice 1)
    "twos" (score-n dice 2)
    "threes" (score-n dice 3)
    "fours" (score-n dice 4)
    "fives" (score-n dice 5)
    "sixes" (score-n dice 6)
    "full house" (score-full-house dice)
    "four of a kind" (score-foak dice)
    "little straight" (score-little-straight dice)
    "big straight" (score-big-straight dice)
    "choice" (reduce + dice)
  )
)
